File: 프로그래머스/연습문제/Lv2/숫자 변환하기/숫자 변환하기.cpp
#include <string>
#include <vector>
#include <set>
using namespace std;

int solution(int x, int y, int n) {
    // 연산 필요 없을 때
    if(x == y) return 0;
    
    set<int> saved_num;
    saved_num.insert(x);
    
    int cnt = 0;
    while(++cnt){
        // 저장된 수들에 3가지 연산을 하기
        set<int> maked_num;
        set<int>::iterator it;
        for(it=saved_num.begin(); it!=saved_num.end(); it++){
            int a = *it+n;
            int b = *it*2;
            int c = *it*3;
            
            if(a==y || b==y || c==y) return cnt;
            else{
                if(a < y) maked_num.insert(a);
                if(b < y) maked_num.insert(b);
                if(c < y) maked_num.insert(c);
            }
        }
        
        saved_num = maked_num;
        if(saved_num.empty()) break;
    }
    
    return -1;
}

File: 프로그래머스/연습문제/Lv2/숫자 변환하기/숫자 변환하기.py
def solution(x, y, n):
    if x == y: return 0

    saved_num = set()
    saved_num.add(x)

    cnt = 0
    while True:
        cnt += 1

        maked_num = set()
        for num in saved_num:
            a = num+n
            b = num*2
            c = num*3

            if a==y or b==y or c==y: return cnt
            else:
                if a < y: maked_num.add(a)
                if b < y: maked_num.add(b)
                if c < y: maked_num.add(c)

        saved_num = maked_num
        if len(saved_num) == 0: break

    return -1
